[doc][llm] Add docs for serving LLMs with TPUs (#65026)
ray-project/ray: doc/source/serve/llm/user-guides/tpu.md

> (serve-llm-tpu)=
> # TPU serving

Reviewer: Every one of the 15 existing pages in `doc/source/serve/llm/user-guides/` carries a `myst.html_meta.description`, and the [style guide](https://docs.ray.io/en/master/ray-contribute/writing-style.html) requires one on every page ("Front matter"). This new page is the only one without it.

````suggestion
---
myst:
  html_meta:
    description: "Serve LLMs on single-host and multi-host TPU slices with Ray Serve LLM, including topology-aware placement groups and a multi-host vLLM TPU example."
---

(serve-llm-tpu)=
# TPU serving
````

> Ray Serve LLM can run a vLLM TPU engine on single-host and multi-host TPU slices. Use this when your Ray cluster already exposes TPU resources and TPU node labels, and your container image includes the TPU variant of vLLM from `tpu-inference`. For Kubernetes setup, see {doc}`Use TPUs with KubeRay </cluster/kubernetes/user-guides/tpu>`.

Reviewer 2: Nit: is the TPUs with KubeRay guide sufficient to satisfy the prerequisites? If so, we could directly instruct people to follow those steps

Reviewer 3: That guide links to https://docs.ray.io/en/latest/cluster/kubernetes/user-guides/gcp-gke-tpu-cluster.html#kuberay-gke-tpu-cluster-setup which does include the full setup on GKE - albeit on older (v4) TPUs. In a separate PR I could update this to at least v6e which is more relevant for serving.

> Ray Serve LLM can run a vLLM TPU engine on single-host and multi-host TPU slices. Use this when your Ray cluster already exposes TPU resources and TPU node labels, and your container image includes the TPU variant of vLLM from `tpu-inference`. For Kubernetes setup, see {doc}`Use TPUs with KubeRay </cluster/kubernetes/user-guides/tpu>`.

Reviewer 2: Could we define "TPU slice"?

Author: Done in https://github.com/ray-project/ray/pull/65026/commits/dd2dab911421e719bfbd43174a6ea9f4c1e31c2b

:::{note}

> TPU support in Ray Serve LLM is topology-aware when you set `accelerator_config={"kind": "tpu", "topology": ...}`. Without a topology, Ray Serve LLM falls back to a regular placement group with per-chip `{"TPU": 1}` bundles.

Reviewer 2: Nit: should we move this lower, once topology-aware has been defined?

Author: Done in https://github.com/ray-project/ray/pull/65026/commits/dd2dab911421e719bfbd43174a6ea9f4c1e31c2b

:::

## Topology and placement

> A TPU topology describes the chip grid in one physical TPU slice. For example, a v6e `4x4` slice has 16 chips. Multi-host slices spread those chips across multiple TPU hosts connected by the TPU interconnect. For v6e `4x4`, that usually means 4 hosts with 4 chips each.

Reviewer 2: could we add a link to https://tpu-visualizer.uc.r.appspot.com/ here? cc: @ryanaoleary, is the app stable?

Author: this animation is cool!

Reviewer 3: Yeah it's linked from the GKE docs: https://docs.cloud.google.com/tpu/docs/system-architecture-tpu-vm#tpu-viz, so it should be good to include.

Author: Done in https://github.com/ray-project/ray/pull/65026/commits/7bff1f5ae39efcc5b5adfeb75fce15288335642f

> Ray Serve LLM uses `tensor_parallel_size * pipeline_parallel_size` as the number of TPU chips requested by one model replica. When you also set a TPU topology, Ray Serve LLM computes the number of chips per host and creates one placement group bundle per TPU host:

Reviewer: Two small things in this span.

Style guide, "Numbers and units": spell out zero through nine in prose, numerals for 10 and above. So "4 hosts with 4 chips each" spells out, while "16 chips" stays a numeral. (`4x4` is a topology literal and stays as-is.)

Style guide, "Write in active voice": "chips requested by one model replica" has a nameable actor, so put it in front of the verb.

```suggestion
A TPU topology describes the chip grid in one physical TPU slice. For example, a v6e `4x4` slice has 16 chips. Multi-host slices spread those chips across multiple TPU hosts connected by the TPU interconnect. For v6e `4x4`, that usually means four hosts with four chips each.

Ray Serve LLM uses `tensor_parallel_size * pipeline_parallel_size` as the number of TPU chips that one model replica requests. When you also set a TPU topology, Ray Serve LLM computes the number of chips per host and creates one placement group bundle per TPU host:
```

```python
LLMConfig(
    accelerator_type="TPU-V6E",
    accelerator_config={"kind": "tpu", "topology": "4x4"},
    engine_kwargs={"tensor_parallel_size": 16},
)
```

> For a v6e `4x4` slice, this produces 4 host-level bundles:

Reviewer: Same spell-out rule as above.

```suggestion
For a v6e `4x4` slice, this produces four host-level bundles:
```

```python
[
    {"TPU": 4, "accelerator_type:TPU-V6E": 0.001},
    {"TPU": 4, "accelerator_type:TPU-V6E": 0.001},
    {"TPU": 4, "accelerator_type:TPU-V6E": 0.001},
    {"TPU": 4, "accelerator_type:TPU-V6E": 0.001},
]
```

> The model still spans all 16 chips because `tensor_parallel_size=16`. The bundle shape only tells Ray how to reserve the hosts that own those chips. If you need per-chip bundles, set `placement_group_config={"bundle_per_worker": {"TPU": 1}}`.

Reviewer 2: Do we know when a user would need per-chip bundles?

Reviewer 3: This is a requirement for Torch backends, which require a single process per TPU chip (or Tensornode).  With vLLM, I believe `TPU_BACKEND_TYPE` currently supports JAX or torchax, but full PyTorch support is planned. I think another possible use-case is single-chip inference.

> 
> ```{figure} ../images/ray_serve_llm_tpu.png
> ---
> width: 100%
> name: ray-serve-llm-tpu-placement
> ---
> Topology-aware TPU placement for one Ray Serve LLM replica.
> ```

Reviewer: Nit, and I'll flag up front that the sibling `{figure}` blocks in `serve/llm/architecture/` don't do this either, so it's a guide rule rather than a local convention. The style guide's "Alt text" section asks for descriptive alt text on every image. The caption renders as visible text below the figure, so it doesn't serve screen readers the way `alt` does.

````suggestion
```{figure} ../images/ray_serve_llm_tpu.png
---
width: 100%
name: ray-serve-llm-tpu-placement
alt: One Ray Serve LLM replica spanning a v6e 4x4 TPU slice, with one placement group bundle per TPU host and four TPU chips reserved in each bundle.
---
Topology-aware TPU placement for one Ray Serve LLM replica.
```
````

Adjust the wording if it doesn't match what the diagram actually shows. I'm reading it from the surrounding prose, not from the image.

Author: Done in https://github.com/ray-project/ray/pull/65026/commits/dd2dab911421e719bfbd43174a6ea9f4c1e31c2b

> ## SlicePlacementGroup

Reviewer: Nit on consistency: the KubeRay TPU guide styles this same identifier as inline code in its heading (`doc/source/cluster/kubernetes/user-guides/tpu.md` line 91). Matching that reads better than a bare identifier used as a title.

```suggestion
## `SlicePlacementGroup`
```

Skip this if @eicherseiji's question below about whether this section is an implementation detail resolves toward cutting it.

For topology-aware TPU configs, Ray Serve LLM creates a {class}`~ray.util.tpu.SlicePlacementGroup` instead of a plain placement group. `SlicePlacementGroup` reserves a matching TPU slice, reads its `ray.io/tpu-slice-name` label, and creates the worker placement group with a per-bundle label selector that pins all bundles to that same physical slice.

> This makes the TPU slice an atomic scheduling unit: a replica reserves the complete slice it needs, or it waits. The placement group doesn't span unrelated slices.

Reviewer: Style guide, "Sentence structure": use a colon only to introduce a list or a code block, and prefer splitting a long sentence over joining clauses with punctuation.

```suggestion
This makes the TPU slice an atomic scheduling unit. A replica reserves the complete slice it needs, or it waits. The placement group doesn't span unrelated slices.
```

> Ray Serve LLM defers this placement group creation until the `LLMServer` replica starts. The server stores the resulting placement group in the vLLM parallel config, and passes it to the TPU vLLM executor.

Reviewer 2: Should we drop this as an implementation detail?

## How the TPU vLLM executor uses the bundles

> The `tpu-inference` Ray executor checks `parallel_config.placement_group`. When Ray Serve LLM already provided one, the executor reuses it instead of creating its own.

Reviewer 2: Could we link to some direct resource about the `tpu-inference` Ray executor?

Author: Done in https://github.com/ray-project/ray/pull/65026/commits/dd2dab911421e719bfbd43174a6ea9f4c1e31c2b

Author: <img width="1293" height="264" alt="Screenshot 2026-08-20 at 11 59 08 AM" src="https://github.com/user-attachments/assets/c48134f5-0a7e-4228-b710-6e8ff0048023" />

Verified the link.

The executor then:

1. Selects the TPU bundles from the placement group.
2. Reads the TPU count from each bundle.

> 3. Starts one Ray worker actor per bundle with `PlacementGroupSchedulingStrategy`, pinning each actor to its bundle.

Reviewer: Two list conventions from the style guide, "Lists and punctuation".

Write list lead-ins as complete sentences ending in a colon. "The executor then:" isn't a complete sentence.

Number every ordered-list item `1.` in Markdown. MyST renders them in source order, so inserting or reordering a step becomes a one-line edit instead of a renumber pass.

```suggestion
The executor then does the following:

1. Selects the TPU bundles from the placement group.
1. Reads the TPU count from each bundle.
1. Starts one Ray worker actor per bundle with `PlacementGroupSchedulingStrategy`, pinning each actor to its bundle.
```

> With the default topology-aware bundles, one vLLM worker actor maps to one TPU host and consumes all local TPU chips on that host. The worker process uses the chips across the multi-host slice. Only if you force `{"TPU": 1}` bundles do you get one worker per chip.

Reviewer 2: ```suggestion
With the default topology-aware bundles, one vLLM worker actor maps to one TPU host and consumes all local TPU chips on that host. The worker process uses the chips across the multi-host slice. Otherwise, manually setting `{"TPU": 1}` results in the creation of one worker per chip.
```

## Example

> Build the image from the vLLM TPU base image and install a Ray wheel that includes TPU topology support. Install the Ray LLM extra without dependencies so pip doesn't replace the base image's TPU vLLM stack with upstream vLLM.

Reviewer 2: > so pip doesn't replace the base image's TPU vLLM stack with upstream vLLM.

This could cause us to miss a legitimate dependency of Ray Serve LLM though, right?

Author: I'll let the dockerfile / python examples below speak for themselves as they are both validated by Ryan.

> Build the image from the vLLM TPU base image and install a Ray wheel that includes TPU topology support. Install the Ray LLM extra without dependencies so pip doesn't replace the base image's TPU vLLM stack with upstream vLLM.

Reviewer 2: ```suggestion
Build the image from the vLLM TPU base image and install a Ray wheel that includes TPU topology support. Install the `ray[llm]` extra without dependencies so pip doesn't replace the base image's TPU vLLM stack with upstream vLLM.
```

::::{tab-set}

:::{tab-item} Dockerfile

```dockerfile
FROM vllm/vllm-tpu:v0.21.0

ENV VLLM_TARGET_DEVICE=tpu
ENV VLLM_XLA_CACHE_PATH=/tmp/vllm_xla_cache
ENV JAX_PLATFORMS=tpu,cpu
ENV TPU_MULTIHOST_BACKEND=ray
ENV TPU_BACKEND_TYPE=jax
ENV ENABLE_PJRT_COMPATIBILITY=true

USER root

# Use a released Ray version or wheel URL that contains TPU topology support.
ARG RAY_PACKAGE="ray"
RUN pip install --no-cache-dir -U "${RAY_PACKAGE}" && \
    pip install --no-cache-dir --no-deps "ray[llm]"

COPY serve_tpu_multihost.py /home/ray/serve_tpu_multihost.py
```

> :::

Author: referred to https://github.com/ray-project/ray/pull/63177#issuecomment-4450391570

Author: @ryanaoleary could you double check the examples in this doc are accurate? hopefully you already have a setup to run these examples easily.

Reviewer 3: Ran the example from this PR on a v6e 4x4 slice successfully:

---

1. Build a a container image based on `vllm/vllm-tpu:nightly` updated with the Ray Nightly wheel to include the required changes:

```dockerfile
FROM vllm/vllm-tpu:nightly

ENV VLLM_TARGET_DEVICE=tpu
ENV VLLM_XLA_CACHE_PATH=/tmp/vllm_xla_cache
ENV JAX_PLATFORMS=tpu,cpu
ENV TPU_MULTIHOST_BACKEND=ray
ENV TPU_BACKEND_TYPE=jax
ENV ENABLE_PJRT_COMPATIBILITY=true

USER root

RUN pip install --no-cache-dir -U "https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-3.0.0.dev0-cp312-cp312-manylinux2014_x86_64.whl" && \
    pip install --no-cache-dir --no-deps "ray[llm]"
```

---

2. Execution Script (`serve_tpu_multihost.py`)

```python
from ray import serve
from ray.serve.llm import LLMConfig, LLMServingArgs, build_openai_app

llm_config = LLMConfig(
    model_loading_config=dict(
        model_id="Qwen/Qwen1.5-0.5B-Chat",
    ),
    accelerator_type="TPU-V6E",
    accelerator_config={"kind": "tpu", "topology": "4x4"},
    engine_kwargs=dict(
        tensor_parallel_size=16,
        max_model_len=2048,
        max_num_batched_tokens=2048,
        distributed_executor_backend="ray",
    ),
)

app = build_openai_app(LLMServingArgs(llm_configs=[llm_config]))

if __name__ == "__main__":
    serve.run(app, blocking=False)
```

---

3. Deployment Logs

```logging
(ServeReplica:default:OpenAiIngress pid=361, ip=10.228.13.14) INFO 07-28 01:08:16 [__init__.py:59] TPU info: node_name=tpu-multihost-group-0 | tpu_type=v6e-16 | worker_id=3 | num_chips=4 | num_cores_per_chip=1
...
(ServeReplica:default:LLMServer:Qwen--Qwen1_5-0_5B-Chat pid=361, ip=10.228.16.14) INFO 2026-07-28 01:08:21,253 default_LLMServer:Qwen--Qwen1_5-0_5B-Chat i4dk5um0 -- Using new placement group <ray.util.placement_group.PlacementGroup object at 0x7d36bef7c260>. {'placement_group_id': '608b5c4a6541895fda73bad089b401000000', 'name': '', 'bundles': {0: {'accelerator_type:TPU-V6E': 0.001, 'CPU': 1.0, 'TPU': 4.0}, 1: {'TPU': 4.0, 'CPU': 1.0, 'accelerator_type:TPU-V6E': 0.001}, 2: {'TPU': 4.0, 'CPU': 1.0, 'accelerator_type:TPU-V6E': 0.001}, 3: {'accelerator_type:TPU-V6E': 0.001, 'CPU': 1.0, 'TPU': 4.0}}, 'bundles_to_node_id': {0: '1e7f28e714c745afcf793bc14fd84fc56b0e8b830c3b13dd923a19f7', 1: '03648c5343a9efef3b8e8957311ffb878a923089a9f074fb46771e05', 2: '5033eedce53942a5fc1149d6446a3c60f7a564196ea598ce56890e5c', 3: '93cde192a63499bbf833e298ab4d1b585a295bd50c9b52c6159c7fc9'}, 'strategy': 'PACK', 'state': 'CREATED', 'stats': {'end_to_end_creation_latency_ms': 2.759, 'scheduling_latency_ms': 2.624, 'scheduling_attempt': 1, 'highest_retry_delay_ms': 0.0, 'scheduling_state': 'FINISHED'}}
...
(_get_vllm_engine_config pid=403, ip=10.228.12.14) INFO 07-28 01:08:41 [model.py:610] Resolved architecture: Qwen2ForCausalLM
(_get_vllm_engine_config pid=403, ip=10.228.12.14) INFO 07-28 01:08:41 [model.py:1744] Using max model len 2048
(_get_vllm_engine_config pid=403, ip=10.228.12.14) INFO 07-28 01:08:41 [scheduler.py:239] Chunked prefill is enabled with max_num_batched_tokens=2048.
(_get_vllm_engine_config pid=403, ip=10.228.12.14) INFO 07-28 01:08:41 [tpu_platform.py:209] Initialized sharding configuration: ShardingConfigManager(total_devices=16, sharding_strategy=ShardingStrategy(tensor_parallelism=16, expert_parallelism=1, sequence_parallelism=1, data_parallelism=1, attention_data_parallelism=1, attention_data_expert_parallelism=1, decode_context_parallelism=1), device_indexes=None)
(_get_vllm_engine_config pid=403, ip=10.228.12.14) INFO 07-28 01:08:41 [tpu_platform.py:321] Force using RayDistributedExecutor for JAX on multihost.
...
(ServeReplica:default:LLMServer:Qwen--Qwen1_5-0_5B-Chat pid=361, ip=10.228.16.14) INFO 2026-07-28 01:08:43,153 default_LLMServer:Qwen--Qwen1_5-0_5B-Chat i4dk5um0 -- Using executor class: <class 'tpu_inference.executors.ray_distributed_executor.RayDistributedExecutor'>
...
(ServeReplica:default:LLMServer:Qwen--Qwen1_5-0_5B-Chat pid=361, ip=10.228.16.14) (EngineCore pid=706) (pid=, ip=10.228.12.14) I0728 01:10:46.806864 4257 deepsea_compiler_backend.cc:2125] XLA::TPU program HBM usage: 918.0K / 31.25G
(ServeReplica:default:LLMServer:Qwen--Qwen1_5-0_5B-Chat pid=361, ip=10.228.16.14) (EngineCore pid=706) (pid=, ip=10.228.12.14) I0728 01:10:46.806864 4257 deepsea_compiler_backend.cc:2158] XLA::TPU program VMEM usage: 55.0K / 128.00M
```

```yaml
$ serve status

proxies:
  18ecab6ff62381ed3acb908975099048d2e5468fa1b155f6e668bcb3: HEALTHY
  5033eedce53942a5fc1149d6446a3c60f7a564196ea598ce56890e5c: HEALTHY
  93cde192a63499bbf833e298ab4d1b585a295bd50c9b52c6159c7fc9: HEALTHY

applications:
  default:
    status: RUNNING
    message: ''
    last_deployed_time_s: 1785200890.5201018
    deployments:
      LLMServer:Qwen--Qwen1_5-0_5B-Chat:
        status: HEALTHY
        status_trigger: CONFIG_UPDATE_COMPLETED
        replica_states:
          RUNNING: 1
        message: ''
      OpenAiIngress:
        status: HEALTHY
        status_trigger: CONFIG_UPDATE_COMPLETED
        replica_states:
          RUNNING: 1
        message: ''

target_capacity: null
```

---

#### Request:
```bash
curl -s http://localhost:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "Qwen/Qwen1.5-0.5B-Chat",
    "messages": [{"role": "user", "content": "Hello! Tell me about TPU v6e in one sentence."}],
    "max_tokens": 50
  }'
```

#### Response:
```json
{
  "id": "chatcmpl-0b20f9fc-be99-4b66-9d92-7b9ef082609a",
  "object": "chat.completion",
  "created": 1785201083,
  "model": "Qwen/Qwen1.5-0.5B-Chat",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "content": "TPU v6e is a model from NVIDIA that supports 64-bit architecture.",
        "refusal": null,
        "annotations": null,
        "audio": null,
        "function_call": null,
        "tool_calls": [],
        "reasoning": null
      },
      "logprobs": null,
      "finish_reason": "stop",
      "stop_reason": null,
      "token_ids": null,
      "routed_experts": null
    }
  ],
  "service_tier": null,
  "system_fingerprint": "vllm-0.22.1rc1.dev192+g063ce98fb-tp16-efa5e651",
  "usage": {
    "prompt_tokens": 33,
    "total_tokens": 52,
    "completion_tokens": 19,
    "prompt_tokens_details": null
  },
  "prompt_logprobs": null,
  "prompt_token_ids": null,
  "prompt_text": null,
  "kv_transfer_params": null
}
```

:::{tab-item} Python

```{literalinclude} ../../../llm/doc_code/serve/tpu/serve_tpu_multihost.py
:language: python
:start-after: __serve_tpu_multihost_start__
:end-before: __serve_tpu_multihost_end__
```
:::

::::

> This example serves `google/gemma-4-31B-it` on one v6e `4x4` slice with TP 16. Set `model_source` to a local or mounted model path, or another path that all TPU hosts can read.

Reviewer 2: Use tool-calling!

> This example serves `google/gemma-4-31B-it` on one v6e `4x4` slice with TP 16. Set `model_source` to a local or mounted model path, or another path that all TPU hosts can read.

Reviewer: "TP 16" is an undefined acronym on first (and only) use. The style guide asks you to spell out the full term first, but here the page already names the parameter, so referring to it directly is shorter than defining an acronym you use once.

```suggestion
This example serves `google/gemma-4-31B-it` on one v6e `4x4` slice with `tensor_parallel_size=16`. Set `model_source` to a local or mounted model path, or another path that all TPU hosts can read.
```

Deliberately not touching the model ID here: @ryanaoleary already confirmed on the Python file that the Gemini bot's claim about it was wrong.

```

:::{note}
TPU support in Ray Serve LLM is topology-aware when you set `accelerator_config={"kind": "tpu", "topology": ...}`. Without a topology, Ray Serve LLM falls back to a regular placement group with per-chip `{"TPU": 1}` bundles.
:::

## `SlicePlacementGroup`

For topology-aware TPU configs, Ray Serve LLM creates a {class}`~ray.util.tpu.SlicePlacementGroup` instead of a plain placement group. `SlicePlacementGroup` reserves a matching TPU slice, reads its `ray.io/tpu-slice-name` label, and creates the worker placement group with a per-bundle label selector that pins all bundles to that same physical slice.

This makes the TPU slice an atomic scheduling unit. A replica reserves the complete slice it needs, and the placement group doesn't span unrelated slices.

## How the TPU vLLM executor uses the bundles

The `tpu-inference` [Ray executor](https://github.com/vllm-project/tpu-inference/blob/main/tpu_inference/executors/ray_distributed_executor_v2.py) checks `parallel_config.placement_group`. When Ray Serve LLM already provided one, the executor reuses it instead of creating its own.

The executor then does the following:

> 1. Selects the TPU bundles from the placement group.
> 1. Reads the TPU count from each bundle.
> 1. Starts one Ray worker actor per bundle with `PlacementGroupSchedulingStrategy`, pinning each actor to its bundle.

Reviewer 2: ```suggestion
1. Selects the TPU bundles from the placement group.
2. Reads the TPU count from each bundle.
3. Starts one Ray worker actor per bundle with `PlacementGroupSchedulingStrategy`, pinning each actor to its bundle.
```

Author: actually, it will render as expected! this is the recommended approach because adding a new numbered item won’t require renumbering the existing ones which keeps the diff minimal.

Author: https://github.com/ray-project/ray/pull/65026#discussion_r3761548047

Author: adopted this pattern and validated in this other PR https://github.com/ray-project/ray/pull/65569.
